### PR TITLE
Make avifImageScale function public

### PR DESCRIPTION
**Pull request by vigneshvg**
This is useful for apps to scale images (without having to
worry about bitdepth, subsampling, etc).

This commit renames avifImageScale to avifImageScaleWithLimit. The
latter would remain an internal only function while the former is
now part of the public ABI (moved to avif.h)

Potential workaround for issue #1484.

### CODE REVIEW

File: include/avif/avif.h [hunk 1]
```diff
@@ -749,6 +749,13 @@ AVIF_API void avifImageStealPlanes(avifImage * dstImage, avifImage * srcImage, a
 // the number of threads that can ever be in flight at a given time, please account for this
 // accordingly.
 
+// ---------------------------------------------------------------------------
+// Scaling
+
+// Scales the YUV/A planes in-place. dstWidth and dstHeight must both be <= AVIF_DEFAULT_IMAGE_DIMENSION_LIMIT and
+// dstWidth*dstHeight should be <= AVIF_DEFAULT_IMAGE_SIZE_LIMIT.
+avifBool avifImageScale(avifImage * image, uint32_t dstWidth, uint32_t dstHeight, avifDiagnostics * diag);
```
Comment (wantehchang) [inline include/avif/avif.h:757]: Public functions should return `avifResult` to indicate success/failure.

Comment (wantehchang) [inline include/avif/avif.h:757]: Please look into whether the public function needs a `diag` parameter. I defer this decision to you.

File: include/avif/avif.h [hunk 2]
```diff
@@ -749,6 +749,13 @@ AVIF_API void avifImageStealPlanes(avifImage * dstImage, avifImage * srcImage, a
 // the number of threads that can ever be in flight at a given time, please account for this
 // accordingly.
 
+// ---------------------------------------------------------------------------
+// Scaling
+
+// Scales the YUV/A planes in-place. dstWidth and dstHeight must both be <= AVIF_DEFAULT_IMAGE_DIMENSION_LIMIT and
+// dstWidth*dstHeight should be <= AVIF_DEFAULT_IMAGE_SIZE_LIMIT.
```
Comment (wantehchang) [inline include/avif/avif.h:756]: I wonder if we can figure out the image dimension and the image size limits of libyuv and use those limits instead.

File: include/avif/avif.h (hunk 1 above)
Comment (y-guyon) [inline include/avif/avif.h:757]: `AVIF_API`

**Review by y-guyon — APPROVED** [review]

**Review by maryla-uc — APPROVED** [review]
Thanks!

File: include/avif/avif.h (hunk 1 above)
Comment (maryla-uc) [reply include/avif/avif.h:757]: I think it doesn't hurt to have it, it's always nice to have more detailed error messages than just a status, or "false". Maybe clarify that it's optional (can be null).

File: src/read.c [hunk 3]
```diff
@@ -4981,13 +4981,13 @@ static avifResult avifDecoderDecodeTiles(avifDecoder * decoder, uint32_t nextIma
...
-                                decoder->imageDimensionLimit,
-                                &decoder->diag)) {
-                avifDiagnosticsPrintf(&decoder->diag, "avifImageScale() failed");
+            if (!avifImageScaleWithLimit(tile->image,
+                                         tile->width,
+                                         tile->height,
+                                         decoder->imageSizeLimit,
+                                         decoder->imageDimensionLimit,
+                                         &decoder->diag)) {
+                avifDiagnosticsPrintf(&decoder->diag, "avifImageScaleWithLimit() failed");
```
Comment (wantehchang) [inline src/read.c:4990]: I think we should delete this `avifDiagnosticsPrintf()` call because it overwrites the better diagnostic message written by `avifImageScaleWithLimit()`. (You can do this in a separate pull request because this is a bug fix.)

File: include/avif/avif.h (hunk 1 above)
Comment (vigneshvg) [reply include/avif/avif.h]: i saw that a few other public non encoder/decoder functions also have to diag parameter. so i thought i was okay to keep.

have clarified that it can be null.

Comment (vigneshvg) [reply include/avif/avif.h]: good point, done.

File: include/avif/avif.h (hunk 2 above)
Comment (vigneshvg) [reply include/avif/avif.h]: existing code uses libavif's internal limits even though it calls libyuv. so i think it is fine.

File: include/avif/avif.h (hunk 1 above)
Comment (vigneshvg) [reply include/avif/avif.h]: Done.

File: src/read.c (hunk 3 above)
Comment (vigneshvg) [reply src/read.c:4990]: Acknowledged.

File: src/scale.c [hunk 4]
```diff
@@ -21,24 +21,24 @@
 // This should be configurable and/or smarter. kFilterBox has the highest quality but is the slowest.
 #define AVIF_LIBYUV_FILTER_MODE kFilterBox
 
-avifBool avifImageScale(avifImage * image,
-                        uint32_t dstWidth,
-                        uint32_t dstHeight,
-                        uint32_t imageSizeLimit,
-                        uint32_t imageDimensionLimit,
-                        avifDiagnostics * diag)
+avifBool avifImageScaleWithLimit(avifImage * image,
```
[the remark this answers is not in the document]
Comment (vigneshvg) [reply src/scale.c]: i don't think there is a need for this yet. we can always add it later.

**Review by y-guyon — APPROVED** [review]

File: include/avif/avif.h [hunk 5]
```diff
@@ -749,6 +749,13 @@ AVIF_API void avifImageStealPlanes(avifImage * dstImage, avifImage * srcImage, a
 // the number of threads that can ever be in flight at a given time, please account for this
 // accordingly.
 
+// ---------------------------------------------------------------------------
+// Scaling
+
+// Scales the YUV/A planes in-place. dstWidth and dstHeight must both be <= AVIF_DEFAULT_IMAGE_DIMENSION_LIMIT and
+// dstWidth*dstHeight should be <= AVIF_DEFAULT_IMAGE_SIZE_LIMIT. diag is optional and can be NULL.
```
Comment (wantehchang) [inline include/avif/avif.h:756]: Remove "diag is optional and can be NULL" for now. Please see my comment on this issue in src/scale.c.

Comment (vigneshvg) [reply include/avif/avif.h:756]: Done.

File: tests/gtest/avifscaletest.cc [hunk 6]
```diff
@@ -50,19 +51,19 @@ TEST_P(ScaleTest, Roundtrip) {
 
   const uint32_t scaled_width = static_cast<uint32_t>(image->width * 0.9);
   const uint32_t scaled_height = static_cast<uint32_t>(image->height * 2.14);
-  const uint32_t kImageSizeLimit = std::numeric_limits<uint32_t>::max();
-  const uint32_t kImageDimensionLimit = std::numeric_limits<uint32_t>::max();
   avifDiagnostics diag;
   avifDiagnosticsClearError(&diag);
-  ASSERT_TRUE(avifImageScale(scaled_image.get(), scaled_width, scaled_height,
-                             kImageSizeLimit, kImageDimensionLimit, &diag))
+  ASSERT_EQ(
```
[the remark this answers is not in the document]
Comment (vigneshvg) [reply tests/gtest/avifscaletest.cc:56]: Done.

File: tests/gtest/avifscaletest.cc [hunk 7]
```diff
@@ -79,6 +80,20 @@ INSTANTIATE_TEST_SUITE_P(
...
             /*create_alpha=*/Values(true, false)));
 
+TEST(ScaleTest, LargerThanDefaultLimits) {
+  const testutil::AvifImagePtr image = testutil::ReadImage(
+      data_path, "paris_exif_xmp_icc.jpg", AVIF_PIXEL_FORMAT_YUV420, 8,
+      AVIF_CHROMA_DOWNSAMPLING_BEST_QUALITY, kIgnoreMetadata, kIgnoreMetadata,
+      kIgnoreMetadata);
+  ASSERT_NE(image, nullptr);
+  // dstWidth*dstHeight is larger than AVIF_DEFAULT_IMAGE_SIZE_LIMIT.
+  EXPECT_NE(avifImageScale(image.get(), 20000, 20000, NULL), AVIF_RESULT_OK);
```
[the remark this answers is not in the document]
Comment (vigneshvg) [reply tests/gtest/avifscaletest.cc:90]: Done.

File: src/scale.c [hunk 8]
```diff
@@ -150,5 +150,10 @@ avifBool avifImageScale(avifImage * image,
         }
     }
 
-    return AVIF_TRUE;
+    return AVIF_RESULT_OK;
+}
+
+avifResult avifImageScale(avifImage * image, uint32_t dstWidth, uint32_t dstHeight, avifDiagnostics * diag)
+{
```
[the remark this answers is not in the document]
Comment (vigneshvg) [reply src/scale.c:157]: Done.

**Comment by vigneshvg** [comment]
@wantehchang i will go ahead and merge this PR. if you have any further comments, i will address them in a follow-up.

File: src/read.c (hunk 3 above)
Comment (vigneshvg) [reply src/read.c:4990]: Done in #1666.
